reasoner/planner/AnswerCountEstimator.java

```diff
@@ -60,210 +59,195 @@
-            this.acyclicConcludables = iterate(resolvables)
-                    .filter(resolvable -> resolvable.isConcludable() && !cyclicConcludables.contains(resolvable.asConcludable()))
-                    .map(Resolvable::asConcludable).toSet();
-        }
+            Map<Variable, CoverElement> cover = new HashMap<>();
+            iterate(variables).forEachRemaining(v -> cover.put(v, new CoverElement(bestUnary.get(v))));
 
-        public FunctionalIterator<Resolvable<?>> resolvables() {
-            return iterate(resolvables);
+            Map<LocalModel, Pair<Set<Variable>, Double>> scaledEstimates = new HashMap<>();
+            relevantModels.forEach(model -> {
+                Set<Variable> modelledVars = intersection(model.variables, variables);
+                scaledEstimates.put(model, new Pair<>(modelledVars, scaledEstimate(model, modelScale.get(model), variables)));
+            });
+            relevantModels.sort(Comparator.comparing(model -> scaledEstimates.get(model).second()));
+            for (LocalModel model : relevantModels) {
+                if (scaledEstimates.get(model).second() < answerEstimateFromCover(scaledEstimates.get(model).first(), cover)) {
+                    CoverElement coverElement = new CoverElement(scaledEstimates.get(model).second()); // Same instance for all keys
+                    scaledEstimates.get(model).first().forEach(v -> cover.put(v, coverElement));
+                }
+            }
+            return answerEstimateFromCover(variables, cover);
         }
 
-        public FunctionalIterator<Retrievable> retrievables() {
-            return iterate(resolvables).filter(Resolvable::isRetrievable).map(Resolvable::asRetrievable);
-        }
+        private static class CoverElement {
+            private final double estimate;
 
-        public FunctionalIterator<Negated> negateds() {
-            return iterate(resolvables).filter(Resolvable::isNegated).map(Resolvable::asNegated);
+            private CoverElement(double estimate) {
+                this.estimate = estimate;
+            }
         }
 
-        public FunctionalIterator<Concludable> cyclicConcludables() {
-            return iterate(cyclicConcludables);
+        private static double scaledEstimate(LocalModel model, Pair<Double, Optional<Variable>> scale, Set<Variable> estimateVariables) {
+            assert scale.second().isPresent() || scale.first() == 1.0;
+            Set<Variable> variables = new HashSet<>();
+            iterate(estimateVariables).filter(model.variables::contains).forEachRemaining(variables::add);
+            double ans = model.estimateAnswers(variables);
+            if (scale.second().isPresent()) {
+                assert scale.first() <= 1.0;
+                variables.add(scale.second().get());
+                double scaledAns = scale.first() * model.estimateAnswers(variables);
+                ans = Math.min(ans, scaledAns);
+            }
+            return ans;
```

Comment:
The effect of the scaling factor on the estimate for a set of variables - 
The scaled estimate is the minimum of 
- The original estimate for the queried set of variables
- The scaling factor * The estimate for the (set of queried variables $$\cup$ the variable which caused the scaling)